include/mata/utils/partition-relation-pair.hh

```diff
@@ -0,0 +1,1449 @@
+*/
+StateBlocks Partition::partition(void)
+{
+    StateBlocks result{};
+    result.insert(result.end(), m_blocks.size(), std::vector<State>());
+    for(auto blockItem : m_blockItems)
+    {
+        result[blockItem.blockIdx].push_back(blockItem.state);
+    }
+    return result;
+}
+
+/** Splitting the blocks of existing partition. According to the input
+* vector of states 'marked', there will be two types of states - marked
+* and unmarked ones. The partition will be split as follows:
+* - if there is a block whose all elements are marked, the block remains
+* unchanged
+* - if there is a block whose all elements are unmarked, the block remains
+* unchanged
+* - if there is a block which contains both marked and unmarked states, it
+* will be split in two blocks such that the first one contains marked states
+* and the second one contains unmarked states of the original block
+* - it means that each block is either unchanged or split in two parts
+* - if a block contains states such that corresponding BlockItems form
+* contiguous subvector on the interval of natural numbers <a, b>, the split
+* nodes will correspond to such BlockItems that they form contiguous subvectors
+* on the intervals of natural numbers <a, k> and <k+1, b>, where a <= k < b.
+* The BlockItems on the interval <a, b> will be swapped such that the property
+* above holds. The representant (first BlockItem on the interval) will always
+* keep its position (the strategy of swapping will adapt to the fact whether
+* the representant is marked or not). Thus, a representant of any node never
+* changes its position.
+* Moreover, the node corresponding to the ancestor of the two split blocks
+* will still describe a valid contiguous interval of natural numbers <a, b>.
+*
+* There are several troubles which can possiby occur:
+* - if an nonexisting state is used, the function detects it and fails
+* - if there is a state which is marked multiple times, the function detects it
+* and fails
+*
+* If a block is split, the function creates a structure SplitPair which
+* contains:
+* - index of the new block which keeps identity of the former block
+* - index of the new block which is newly constructed
+* - index of the node which is an ancestor of these two blocks
+* The function returns a vector of such SplitPairs.
+*
+* @brief splits blocks of the partition
+* @param marked marked states which influence splitting
+* @return vector of SplitPair which contains information about split blocks
+*/
+std::vector<SplitPair> Partition::splitBlocks(const std::vector<State>& marked)
+{
+    // the vector which will be returned as the result
+    std::vector<SplitPair> split{};
+    
+    // if there is no marked state, no block could be split
+    if(marked.empty()) { return split; }
+    
+    // this vector contains information about states which has been marked
+    // and helps to detect states which has been marked multiple times
+    std::vector<char> usedStates{};
```

Comment:
@Adda0 Thanks for your suggestion. I tried to avoid using `std::vector<bool>` in this implementation from the very beginning since it does not ensure data locality. Instead, I decided to use `std::vector<char>` even for boolean values even it is not as space-efficient because it is a contiguous block of memory. I discussed it with Lukáš and he agreed.

Reply:
Sounds good to me. Closing this conversation.